module-1/lab-errhand_listcomp/main.py

```diff
@@ -0,0 +1,385 @@
+#!/usr/bin/env python
```

Comment:
Question 5: really good and simple approach to get odd numbers. This approach make your code more readable for  others: 
```Python
odds = [num for num in range(1,100) if num % 2 != 0]
print(odds)
```
I encourage you to write down at least 2 different ways of solving short questions (future exercises). It's a good way of practising code skills. 
Question 6: really good approach. To make it more readable include a conditional:
```Python
divisible_by_seven= [num for num in range(1,1000) if num % 7 == 0]
print(divisible_by_seven)
```
Question 7: You don't need to use join to solve this problem. A way of doing it, the one shown in class on Friday: 
```Python
 teststring = 'Find all of the words in a string that are monosyllabic'

vowels = ['a','e','i','o','u',' ']
non_vowels = [letter for letter in teststring if letter.lower() not in vowels]
print(non_vowels)
```
Question 10: Good approach!! You need to include madrid-oct-2018 as requested otherwise your approach is throwing your root directory (your first attempt). Remember to use always relative paths instead of absolute ones as in your other two approaches. 
Question 12: This is a for loop , try to convert it to a list comprehension as requested in the exercise.  
```Python
list_of_lists = [[1,2,3],[4,5,6],[7,8,9]]
flatten_list = [y for x in list_of_lists for y in x]
print(flatten_list)
#AlbertoG's approach using numpy flatten function
import numpy as np
flatten_list = np.array(list_of_lists).flatten()
```
Question 13: Try to convert your for loop to a list comprehension as requested in the question. 
Question 14: You have a syntax error because you have missed parenthesis in a print statement. Try to convert your for loop to a list comprehension as requested in the exercise. 
Question 15: you were expected to throw an exception of ZeroDivisionError. An example could be: 
```Python
x = 5
y = 0

try:
    z = x/y
except ZeroDivisionError:
    print ("Can't divide by Zero!")
finally:
    print ('All Done!')
```
Question 16: the same as the former question you were expected to raise an IndexError type. 
```Python

abc=[10,20,20]

#print(abc[3])
try:
    i = int(input("Enter item index to print: "))
except IndexError:
    print ("List index out of range!")
else:
    print(abc[i])
```
Question 17: Good approach but you missed to include try/except block in your code as requested. A possible approach could be: 
```Python

try:
    x = int( input( "enter a number: " ) )
    y = int( input( "enter another number: " ) )
    print( x, '/', y, '=', x/y )
except ZeroDivisionError:
    print( "Can't divide by 0!" )
except ValueError:
    print( "That doesn't look like a number!" )
except:
    print( "something unexpected happend!" )
```
Questions 18 and 19: you were expected to raise specific errors (IOError and ValueError). A possible approach: 
```Python
#18. Handle the exception thrown by the code below by using try and except blocks. 
# Check in provided resources the type of error you may use. 

try:
    f = open('test','r')
    f.write('Test write this')
except IOError:
   print ("Error: File not found or is read-only")
else:
   print ("Content written successfully")
   f.close()


#19. Handle the exceptions that can be thrown by the code below using try and except blocks. 
#Hint: the file could not exist and the data could not be convertable to int

try:
    fp = open('myfile.txt')
    line = f.readline()
    i = int(s.strip())
except (IOError,ValueError) as e:
    print ("Please check the file,either the file is read-only or the data can't be converted to an integer.",e.errno)
except: 
    print ("Unexpected error")
```
Question 20: you were expected to raise an AssertionError and include try/except block/s in your code. 
```Python
def linux_interaction():
    assert ('linux' in sys.platform), "Function can only run on Linux systems."
    print('Doing something.')

try:
    linux_interaction()
except AssertionError as error:
    print(error)
    print('The linux_interaction() function was not executed')
```